pkg/channel/consolidated/dispatcher/dispatcher.go

```diff
@@ -270,26 +274,38 @@ func (d *KafkaDispatcher) UpdateKafkaConsumers(config *Config) (map[types.UID]er
 	d.logger.Debug("Number of subs failed to subscribe", zap.Any("subs", len(failedToSubscribe)))
 
 	// Unsubscribe and close consumer for any deleted subscriptions
-	for channelRef, subs := range d.channelSubscriptions {
-		for _, oldSub := range subs {
-			removedSub := true
-			for _, s := range newSubs {
-				if s == oldSub {
-					removedSub = false
-				}
-			}
+	subsToRemove := make(map[eventingchannels.ChannelReference][]types.UID)
+	for channelRef, actualSubs := range d.channelSubscriptions {
+		subsToRemove[channelRef] = uidSetDifference(actualSubs, newSubs)
+	}
 
-			if removedSub {
-				if err := d.unsubscribe(channelRef, d.subscriptions[oldSub]); err != nil {
-					return nil, err
-				}
+	for channelRef, subs := range subsToRemove {
+		for _, s := range subs {
+			if err := d.unsubscribe(channelRef, d.subscriptions[s]); err != nil {
+				return nil, err
 			}
 		}
 		d.channelSubscriptions[channelRef] = newSubs
 	}
+
 	return failedToSubscribe, nil
 }
 
+func uidSetDifference(a, b []types.UID) (diff []types.UID) {
+	m := make(map[types.UID]bool)
+
+	for _, item := range b {
+		m[item] = true
+	}
+
+	for _, item := range a {
+		if _, ok := m[item]; !ok {
+			diff = append(diff, item)
+		}
+	}
+	return
```

Comment:
Why this?

Reply:
ah ok fine

Reply:
I ran `gcp dd8077f5887074044a03e0d864460aab02aca86a` for the 45 backport

and `gcp 665a13a478217a1d1197859d07855b1daec486e2` for the second backport